clients/src/main/java/org/apache/kafka/common/requests/DescribeAclsRequest.java

```diff
@@ -18,117 +18,114 @@
-import static org.apache.kafka.common.protocol.CommonFields.OPERATION;
-import static org.apache.kafka.common.protocol.CommonFields.PERMISSION_TYPE;
-import static org.apache.kafka.common.protocol.CommonFields.PRINCIPAL_FILTER;
-import static org.apache.kafka.common.protocol.CommonFields.RESOURCE_NAME_FILTER;
-import static org.apache.kafka.common.protocol.CommonFields.RESOURCE_PATTERN_TYPE_FILTER;
-import static org.apache.kafka.common.protocol.CommonFields.RESOURCE_TYPE;
 
 public class DescribeAclsRequest extends AbstractRequest {
-    private static final Schema DESCRIBE_ACLS_REQUEST_V0 = new Schema(
-            RESOURCE_TYPE,
-            RESOURCE_NAME_FILTER,
-            PRINCIPAL_FILTER,
-            HOST_FILTER,
-            OPERATION,
-            PERMISSION_TYPE);
-
-    /**
-     * V1 sees a new `RESOURCE_PATTERN_TYPE_FILTER` that controls how the filter handles different resource pattern types.
-     * For more info, see {@link PatternType}.
-     *
-     * Also, when the quota is violated, brokers will respond to a version 1 or later request before throttling.
-     */
-    private static final Schema DESCRIBE_ACLS_REQUEST_V1 = new Schema(
-            RESOURCE_TYPE,
-            RESOURCE_NAME_FILTER,
-            RESOURCE_PATTERN_TYPE_FILTER,
-            PRINCIPAL_FILTER,
-            HOST_FILTER,
-            OPERATION,
-            PERMISSION_TYPE);
-
-    public static Schema[] schemaVersions() {
-        return new Schema[]{DESCRIBE_ACLS_REQUEST_V0, DESCRIBE_ACLS_REQUEST_V1};
-    }
 
     public static class Builder extends AbstractRequest.Builder<DescribeAclsRequest> {
-        private final AclBindingFilter filter;
+        private final DescribeAclsRequestData data;
 
         public Builder(AclBindingFilter filter) {
             super(ApiKeys.DESCRIBE_ACLS);
-            this.filter = filter;
+            ResourcePatternFilter patternFilter = filter.patternFilter();
+            AccessControlEntryFilter entryFilter = filter.entryFilter();
+            data = new DescribeAclsRequestData()
+                .setHostFilter(entryFilter.host())
+                .setOperation(entryFilter.operation().code())
+                .setPermissionType(entryFilter.permissionType().code())
+                .setPrincipalFilter(entryFilter.principal())
+                .setResourceNameFilter(patternFilter.name())
+                .setResourcePatternType(patternFilter.patternType().code())
+                .setResourceType(patternFilter.resourceType().code());
         }
 
         @Override
         public DescribeAclsRequest build(short version) {
-            return new DescribeAclsRequest(filter, version);
+            validate(version);
+            return new DescribeAclsRequest(data, version);
         }
 
         @Override
         public String toString() {
-            return "(type=DescribeAclsRequest, filter=" + filter + ")";
+            return data.toString();
+        }
+
+        private void validate(short version) {
```

Comment:
Why are we validating in the builder instead of the request?

Reply:
I think it's important to be able to create bogus Requests/Responses in order to validate error handling on the broker side. By validating in the Builder, we allow creating arbitrary objects via the constructor. I expect all callers to use the Builder. 

This is what a few requests already do (`CreateTopicsRequest`, `FindCoordinatorRequest`, `HeartbeatRequest`) and I liked that pattern.

Reply:
I would say it's important  _not_ to be able to create bogus requests. ;) We can introduce specific mechanisms for testing, but a public constructor for a request should do its own validation.

Reply:
Right! 👍